### PR TITLE
[4.2] Move Show operation setup in its own method

### PR DESCRIPTION
**Pull request by tabacitu**
This PR:
- fixes https://github.com/Laravel-Backpack/CRUD/issues/2051
- fixes https://github.com/Laravel-Backpack/CRUD/issues/3108

# Context

In Backpack 4.0 and 4.1, in order to keep backwards-compatibility, we've kept a behaviour for the Show operation that is not intuitive at all. 

The operation tries to guess which columns to show, if none were provided (which is great), but because it does that, it makes it difficult for developer to remove columns from the Show operation. 

# What does this do

This commit:
- fixes that, by moving that logic inside `autoSetupShowOperation()`, which is automatically called if `setupShowOperation()` does not exist;
- opens up the opportunity for developers to completely remove the default behaviour (by creating a `setupShowOperation()` method in their controllers);
- opens up the opportunity for developers to do stuff before or after that bit of logic (by calling `autoSetupShowOperation()` wherever they want inside their `setupShowOperation()`);

# Notes

**THIS IS A BREAKING CHANGE. We can only include this in Backpack 4.2.**

# Todos

Also needs:
- [x] the List operation documentation to be updated
- [x] the 4.2 upgrade guide should instruct people 
     - [x] to change their `setupShowOperation()` methods to call `autoSetupShowOperation()` - that should work well in most use cases;
     - [x] behaviour change - `created_at` and `updated_at` columns will be added if there;
     - [x] behaviour change - `deleted_at` column will be added if model has `SoftDeletes` (and page will be visible for SoftDeleted items)

> Step X. **The default behaviour for the Show operation, in Backpack 4.1, has been to guess which columns to show.** You could then create `setupShowOperation()` to add or modify columns, but you couldn't _remove_ columns, because the default logic would add them back. In Backpack 4.2 we've fixed this, by the moving the "automatic setup" logic inside a `autoSetupShowOperation()` method, that is no longer called if you've defined `setupShowOperation()`. This means 
> - that _if you have a `setupShowOperation()` method in your CrudControllers_, the default guessing of columns will be _completely removed_, you have complete control over what columns are shown; 
> - **To keep the same behaviour as before (guess columns), please call `autoSetupShowOperation()` wherever you want inside your `setupShowOperation()`.**
> - If you do not have a `setupShowOperation()` in your CrudControllers, you are not affected by this, don't worry;

### CODE REVIEW

**Comment by tabacitu** [comment]
Hmmm you're right @mahasadhu - I see we're still removing relationship columns that have been `autoset`... but... I don't think we need to do that any more. AFAIK they should work just fine, since Backpack now tries to guess all missing relationship info. So I believe we can remove this bit:

```php
            // remove any autoset relationship columns
            if (array_key_exists('model', $column) && array_key_exists('autoset', $column) && $column['autoset']) {
                $this->crud->removeColumn($column['key']);
            }
```

Could you please confirm that works for you then?

Of course... this _would_ change the default behaviour of the operation... in that it also shows all relationships by default... but I think that's more intuitive than what we had before anyway... right?

**Comment by mahasadhu** [comment]
@tabacitu - I have checked and yes, it works. However I don't know how other people is using backpack, so I'm pretty sure it's going to break some people's apps.

By the way I don't really understand what is autoset. Is that something that's been used in backpack v3 ?
Because I cannot find autoset mentioned in 4.1 docs. Or is it just me that don't see it. 😕

**Comment by tabacitu** [comment]
Yeah... I don't think _it is_ documented 🙄 ... 

`autoset` is a flag added to columns/fields that have been added using `setFromDb()`. It's there because usually we can't trust things added by `setFromDb()` - at least that's what we thought in Backpack 3.x and 4.0. So we needed a way to identify which fields/columns have been added like that, so we don't trust their definition 100%. 

But. Starting with Backpack 4.1, we have intelligent guessing of field/column attributes, which is _another_ layer of attribute guessing. This one is a lot more reliable, and it's done whenever someone does `addField()` or `addColumn()`. Basically most of the time you only need to give the `name` of a field/column, and the rest Backpack tries to guess itself.

The plan is to phase out `setFromDb()`, along with its `autoset` flag, and only keep the new intelligent guessing.
Because you don't need to do `setFromDb()` anymore, you just need to do `addField()`/`addColumn()` for all the columns in the db (and relationships), and just specify the `name`. Every other attribute, Backpack will guess.

**TLDR: `autoset` is a flag added by `setFromDb()`.**

**Comment by mahasadhu** [comment]
@tabacitu - I see... so that how it should be used.

Currently I'm using `setFromDb()` because that's what I saw on docs. And then I modify any columns that I need using `modifyColumn()` or `modifyField()`. 

**But then again**, sometimes `modifyColumn()` or `modifyField()` does not work (not sure why), and I'm simply call `removeField()` then `addField()` right after it to make it works.

And today I just realized, that looks like `setFromDb()` executes query from information schema. And that query becomes the longest one to run / execute.
image

Looks like there are quite some things that I need to update from my code 🤕 

-------------------------------------------------------------
By the way today I tried to run `DB::getQueryLog()` on the search method in Backpack. And it seems like there are some things that could be optimized. For example this one below, there is one exact same query without any parameters that executed twice.
image

It looks like being used to show total number of record / row in that particular table. It looks like the problem is around here
https://github.com/Laravel-Backpack/CRUD/blob/5414225a0040f2b8b18eb9e3a55d613bfb8d3db6/src/app/Http/Controllers/Operations/ListOperation.php#L76-L77
**But still**, I'm not sure whether this is my mistake, laravel / eloquent mistake, or actually a bug inside the Backpack.

This last bit could be in a different issue though.

**Comment by tabacitu** [comment]
> And today I just realized, that looks like setFromDb() executes query from information schema. And that query becomes the longest one to run / execute.

Yeah, that's just _one_ of the reasons we've stopped recommending `setFromDb()`. The other being that it's IMPOSSIBLE for us to push changes to it, if people use it in production. Any change we make will make thousands of admin panels behave differently than what they did before the change. Not good. 

Currently the recommended way to use Backpack is the one in the [video course](https://backpackforlaravel.com/docs/4.1/getting-started-videos), which is also the one that our generators use - the [new fluent syntax in Backpack 4.1](https://backpackforlaravel.com/docs/4.1/release-notes#alternative-fluent-syntax-for-fields-columns-filters-buttons). But I just realised - we do NOT teach/recommend the same fluent syntax in the text tutorials - so thank you, I've just [made a note to fix this](https://github.com/Laravel-Backpack/docs/issues/173).

> By the way today I tried to run DB::getQueryLog() on the search method in Backpack. And it seems like there are some things that could be optimized. For example this one below, there is one exact same query without any parameters that executed twice.

Indeed, there might be a few things we can optimize in terms of queries on the List operation. We're open to suggestions. The biggest optimization we can make, though, is getting rid of DataTables. It has served us VERY well for a number of years, but now it's holding us back - difficult/impossible to support billions of rows (because it still needs pagination numbers), difficult o make one row a certain color. In a future version we plan to code something from scratch for this operation, probably using Livewire. But until then - optimization are welcome and very appreciated.

I agree - we've diverged enough, let's keep this issue about the Show operation and open other Issues/Discussions for other stuff.

**Comment by tabacitu** [comment]
Fixed conflicts, tested and added [docs for it here](https://github.com/Laravel-Backpack/docs/pull/285). I think it's ready and I want to merge this.

@pxpm & @promatik could you please BOTH review this, and make sure 
- [ ] it works
- [ ] the way `setupShowOperation()` now works is intuitive
- [ ] the way `modifyShowOperation()` now works is intuitive
- [ ] the upgrade guide makes sense and is easy to follow
- [ ] there's no huge negative implication that I missed

Thank youuu 🙏

File: src/app/Http/Controllers/Operations/ShowOperation.php [hunk 1]
```diff
@@ -70,18 +70,43 @@ public function show($id)
...
+        // guess which columns to show, from the database table
+        if ($this->crud->get('show.setFromDb')) {
             $this->crud->setFromDb(false, true);
         }
 
+        // if the developer has chosen to include a modifyShowOperation() method
+        // in his/her CrudController, make sure to call it because they probably want to
+        // add a new column or remove an autoset column
+        if (method_exists($this, 'modifyShowOperation')) {
+            $this->modifyShowOperation();
```
Comment (pxpm) [inline src/app/Http/Controllers/Operations/ShowOperation.php:98]: We don't have this behavior in any other operations, and I think it also does not make sense since the `setupShowOperation` could now be defined by developer. 

My rationale is: if you write `public function modifyShowOperation()` why not `public function setupShowOperation()`? 

What I guess will happen is:

```php
public function setupShowOperation() {
     $this->setupListOperation();
}
```
What am missing here? 

Pedro

Comment (tabacitu) [reply src/app/Http/Controllers/Operations/ShowOperation.php:98]: Good catch. This inconsistency with other operations is something that I feel uncomfortable about too. And that's what I expect most _new_ people to do too - call `$this->setupListOperation()`.

But there's quite a big difference between that and what `modifyShowOperation()` does. Say you have 4 columns added to `list`, out of total 10 columns in the db table. Then:
- if you do nothing, you don't define `setupShowOperation()`, you'll have all 10 columns in your show;
- if you do `setupShowOperation() { $this->setupListOperation(); }` you'll have 4 columns in your Show; those other ones... are lost, you have to define them manually or call `$this->crud->setFromDb()` yourself;
- if you do `modifyShowOperation() {}`, inside there all 10 columns have been added, you can change them to different types or remove them; 

Again, I'm not very happy with this different behaviour either. But it's what I want most of the time from the Show operation - show me not only the columns in the List, but all the columns in the db. Then, what I want is to change column types. That's how I find myself using it.

One way to look at it is that:
- because List is limited in space (each column takes up width), by default no columns are present;
- because Show is unlimited in space (you can easily scroll down), by default all columns are present;

That's the way it's functioned so far, the default behaviour for Show has been to show all columns in the db. We _could_ change this default behaviour, but... I don't think we should. It's useful. It gives you 90% of what you want, most of the time.

Comment (pxpm) [reply src/app/Http/Controllers/Operations/ShowOperation.php:98]: Let me give this another try, I think I got where you are going, but I think I didn't explain myself perfectly clear: 

I still think that method is not needed. If you add columns they don't delete the previous columns, atmost it overrides the ones with same names (or it should). 

I think if it's "default" show operation behaviour, it should be in `setupShowOperationDefaults`, leaving the `setupShowOperation` to be fully controled by developer. 

Wouldn't that make it similar in behavior with all other operations and still keep the ability to "modify" whatever you want inside `setupShowOperation` ? 

Let me know,
Pedro

Comment (tabacitu) [reply src/app/Http/Controllers/Operations/ShowOperation.php:98]: I agree - it would make A LOT more sense in `setupShowOperationDefaults()`, in terms of where code goes. That would be so clean! But the problem is... the order they're run in:
- first `setupShowOperationDefaults()`
- then `setupShowOperation()`

That means if we do all the adding and removing of columns in `setupShowOperationDefaults()`... there's no way for the dev to easily turn that off, from `setupShowOperation()`. It's kind of an irreversible behaviour. 

Even if we do it `setupShowOperationDefaults()` in a conditional (`if (CRUD::get('show.autoset')) {}`)... the dev cannot change that config _before_ everything happens... only afterwards...

If I remember correctly, that's the reason I ended up with `modifyShowOperation()`... that and (of course) upgradability.

Comment (pxpm) [reply src/app/Http/Controllers/Operations/ShowOperation.php:98]: Isn't possible: 
```php
if(method_exists(setupShowOperation)) { dont run please } ?
```
We just need to tell devs that want to customize the operation to manually call their initial setup ? 

```php
public function setupShowOperation() {
     $this->setupShowOperationDefaults();
}
```
We are telling them to use `modifyShowOperation` if they want to achieve the same: - use default, modify after so ... why just don't tell them a different text, but that makes more sense, atleast for me:
- NO setupOperation: setupDefaults is run.
- setupOperation: Defaults are not run if not specified by dev because they are taking control of setup.

```php
public function setupShowOperation() {
    $this->setupShowOperationDefaults();
    $this->removeColumn('xxx');
}
```

If you are worried about "consistency" about operation, we can introduce an operation feature that is: `run defaults on load or allways run defaults`, so operations like create etc run the defaults even when `setupCreateOperation` is present. But it would benefit operations like this one, that would have `run defaults on load => false` that would make it not run defaults when developer takes control of operation setup. 

PS: forget the name, even I reading this see it's completely wrong, but I hope you got my idead.

Comment (tabacitu) [reply src/app/Http/Controllers/Operations/ShowOperation.php:98]: AAHAHAHAH!!! 😅

stupid_me_3

You're totally right @pxpm - let me try it out, we should be able to use `if(method_exists(setupShowOperation))` for this. We can't call `setupShowOperationDefaults()` because that does extra things too, and will be called no matter what, but we can just as well create a new function with the default logic, and call that. So it could be something like:

```php
// no setupShowOperation() defined
// setupShowOperationDefaults() will run guessColumnsForShowOperation()

// otherwise
public function setupShowOperation() {
    $this->guessColumnsForShowOperation();
    $this->removeColumn('xxx');
}
```

For the upgrader then, it's just as easy. We don't tell them "change `setupShowOperation()` to `modifyShowOperation()`, we tell them:

```diff
public function setupShowOperation() {
+    $this->guessColumnsForShowOperation();
      // your existing logic
```

Which is just as good (if not better). I LOVE IT. Thanks a lot @pxpm , great idea!

Comment (tabacitu) [reply src/app/Http/Controllers/Operations/ShowOperation.php]: Done! Changed it, great catch @pxpm .

**Comment by tabacitu** [comment]
@pxpm I've just removed `modifyShowOperation()` like you suggested. Indeed it wasn't needed at all. Good catch!

But we still needed a way to call the "_default column-adding logic_" (aka magic) inside `setupShowOperation()`, and I think I've found a good name for it. So here's how it works now:
- if you don't have a `setupShowOperation()` defined, Backpack will try guess columns;
- if you do have `setupShowOperation()` defined, no magic is performed, you have zero columns added; but inside `setupShowOperation()` you can now call `$this->autoSetupShowOperation()`, which will do all the magic; 

This way, you can do stuff before & after the magic, by doing stuff before/after the `$this->autoSetupShowOperation();` call. But you'll probably mostly do stuff _after_ it - things like adding columns and removing columns. Which was not possible before, you could only do stuff _before_ the magic.

This also makes it easy to upgrade:
- you want the columns autosetup? add `$this->autoSetupShowOperation()` inside your `setupShowOperation()` (at the end for no change in behavior);
- don't want it? don't call it then, and add whatever columns you want inside `setupShowOperation()`, it's a clean slate;

I've updated the upgrade guide & Show operation docs with this behavior. 

**Let me know if something triggers your spidey-sense... or I can merge this.**

Cheers!

PS. I've also merged two more features inside this PR, even if unrelated. But they were easy ones:
- https://github.com/Laravel-Backpack/CRUD/pull/4006
- https://github.com/Laravel-Backpack/CRUD/pull/4007

They just add the `created_at`, `updated_at` and `deleted_at` columns to the Show operation, by default, if config value isn't set otherwise. I think that's a much better default, and it will make it so less people have to define a `setupShowOperation()` in the first place.

File: src/app/Http/Controllers/Operations/ShowOperation.php [hunk 2]
```diff
@@ -113,13 +148,7 @@ public function show($id)
             }
         }
 
-        // remove preview button from stack:line
-        $this->crud->removeButton('show');
-
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ShowOperation.php:118]: This was no longer needed, since in 4.1 we only add that button to the List operation.

File: src/config/backpack/operations/show.php [hunk 3]
```diff
@@ -10,4 +10,15 @@
...
+
+    // Automatically add all columns from the db table?
+    'setFromDb'  => true,
+
+    // Automatically add created_at and updated_at columns, if model has timestamps?
+    'timestamps' => true,
+
+    // If model has SoftDeletes, allow the admin to access the Show page for
+    // soft deleted items & add a deleted_at column to ShowOperation?
+    'softDeletes' => true,
```
Comment (tabacitu) [inline src/config/backpack/operations/show.php:22]:
```suggestion
    'softDeletes' => false,
```
